src/linker/Linker.Steps/TypeMapStep.cs

```diff
@@ -137,6 +158,53 @@ static MethodDefinition GetBaseMethodInTypeHierarchy (TypeDefinition type, Metho
 			return null;
 		}
 
+		// Returns a list of default implementations of the given interface method on this type.
+		// Note that this returns a list to potentially cover the diamond case (more than one
+		// most specific implementation of the given interface methods). Linker needs to preserve
+		// all the implementations so that the proper exception can be thrown at runtime.
+		static IEnumerable<InterfaceImplementation> GetDefaultInterfaceImplementations (TypeDefinition type, MethodDefinition interfaceMethod)
+		{
+			// Go over all interfaces, trying to find a method that is an explicit MethodImpl of the
+			// interface method in question.
+			foreach (var interfaceImpl in type.Interfaces) {
+				var potentialImplInterface = interfaceImpl.InterfaceType.Resolve ();
+				if (potentialImplInterface == null)
+					continue;
+
+				bool foundImpl = false;
+
+				foreach (var potentialImplMethod in potentialImplInterface.Methods) {
+					if (potentialImplMethod == interfaceMethod &&
+						!potentialImplMethod.IsAbstract) {
+						yield return interfaceImpl;
+					}
+
+					if (!potentialImplMethod.HasOverrides)
+						continue;
+
+					// This method is an override of something. Let's see if it's the method we are looking for.
+					foreach (var @override in potentialImplMethod.Overrides) {
+						if (@override.Resolve () == interfaceMethod) {
+							yield return interfaceImpl;
```

Comment:
Just curious: Is it guaranteed that the override is not an abstract method?

Reply:
It could be abstract. I think this is the C# syntax for it:

```csharp
interface IFoo
{
    void Frob() { Console.WriteLine("Hi"); }
}

interface IBar : IFoo
{
    abstract void IFoo.Frob();
}
```

Reply:
If we filter them out, then we would "fix" following code:

```
interface IFoo
{
    void Frob() { Console.WriteLine("Hi"); }
}

interface IBar : IFoo
{
    abstract void IFoo.Frob();
}

class Program : IBar
{
    static void Main() => ((IFoo)new Program()).Frob();
}
```

Before linking, we would get an exception because the default implementation got reabstracted and it's no longer available in IBar. After linking, linker would strip IBar and IBar.IFoo.Frob out as "unused" and make the program work. I don't think that's desirable.

Reply:
(Of course this is only testable from IL, or by compiling into multiple assemblies and switching them around before running. The C# compiler will not allow generation of an "incomplete" type.)